controllers/apim/ratelimitpolicy_controller.go

```diff
@@ -85,15 +100,343 @@ func (r *RateLimitPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl
+			Context: networkingv1alpha3.EnvoyFilter_GATEWAY,
+			ObjectTypes: &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectMatch_Listener{
+				Listener: &networkingv1alpha3.EnvoyFilter_ListenerMatch{
+					PortNumber: 8080, // Kuadrant-gateway listens on this port by default
+					FilterChain: &networkingv1alpha3.EnvoyFilter_ListenerMatch_FilterChainMatch{
+						Filter: &networkingv1alpha3.EnvoyFilter_ListenerMatch_FilterMatch{
+							Name: "envoy.filters.network.http_connection_manager",
+							SubFilter: &networkingv1alpha3.EnvoyFilter_ListenerMatch_SubFilterMatch{
+								Name: "envoy.filters.http.router",
+							},
+						},
+					},
+				},
+			},
+		},
+		Patch: &prePatch,
+	}
+
+	postAuthFilterPatch := preAuthFilterPatch
+	postAuthFilterPatch.Patch = postPatch
+
+	// Eventually, this should be dropped since it's a temp-fix for Kuadrant/limitador#53
+	clusterPatch := networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectPatch{
+		ApplyTo: networkingv1alpha3.EnvoyFilter_CLUSTER,
+		Match: &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectMatch{
+			Context: networkingv1alpha3.EnvoyFilter_GATEWAY,
+			ObjectTypes: &networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectMatch_Cluster{
+				Cluster: &networkingv1alpha3.EnvoyFilter_ClusterMatch{
+					Name: DefaultLimitadorClusterName,
+				},
+			},
+		},
+		Patch: &networkingv1alpha3.EnvoyFilter_Patch{
+			Operation: networkingv1alpha3.EnvoyFilter_Patch_MERGE,
+			Value: &types.Struct{
+				Fields: map[string]*types.Value{
+					"name": {
+						Kind: &types.Value_StringValue{
+							StringValue: PatchedLimitadorClusterName,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	factory := istioprovider.EnvoyFilterFactory{
+		ObjectName: gateway.Name + "-ratelimit-filters",
+		Namespace:  gateway.Namespace,
+		Patches: []*networkingv1alpha3.EnvoyFilter_EnvoyConfigObjectPatch{
+			// Ordering matters here!
+			&preAuthFilterPatch,
+			&postAuthFilterPatch,
+			&clusterPatch,
+		},
+	}
+
+	return factory.EnvoyFilter()
+}
+
+func routeRateLimitsPatch(vHostName, routeName string, actions []*apimv1alpha1.Action_Specifier, stage apimv1alpha1.RateLimit_Stage) *istio.EnvoyFilter {
+	stageValue := apimv1alpha1.RateLimit_Stage_value[apimv1alpha1.RateLimitStage_PREAUTH]
+	if stage == apimv1alpha1.RateLimitStage_POSTAUTH {
+		stageValue = apimv1alpha1.RateLimit_Stage_value[apimv1alpha1.RateLimitStage_POSTAUTH]
+	}
+
+	patchUnstructured := map[string]interface{}{
+		"operation": "MERGE",
+		"value": map[string]interface{}{
+			"route": map[string]interface{}{
+				"rate_limits": []map[string]interface{}{
+					{
+						"stage":   stageValue,
+						"actions": "ACTIONS", // this is replaced with actual value below
+					},
+				},
+			},
+		},
+	}
+
+	patchRaw, _ := json.Marshal(patchUnstructured)
+	stringPatch := string(patchRaw)
+
+	jsonActions, _ := json.Marshal(actions)
+	// A nice trick to make it easier add-in actions
+	stringPatch = strings.Replace(stringPatch, "\"ACTIONS\"", string(jsonActions), 1)
```

Comment:
This should not be done. Structured data should not be serialized to apply string substitution and then de-serialize 
again. 

You can build the unstructured (based on maps and interfaces) actions list object and then apply to the `patchUnstructured` object.


For now it is ok

Reply:
perhaps we should create a TODO or issue to tack the need to change this?

Reply:
For some reason, deserializing actions into that unstructured patch was working properly but I agree this needs a better way. Will create an issue when the PR is merged.